File: app/src/main/java/org/libreoffice/kit/DirectBufferAllocator.java


package org.libreoffice.kit;

import java.nio.ByteBuffer;

public final class DirectBufferAllocator {

    private static final String LOGTAG = DirectBufferAllocator.class.getSimpleName();

    private DirectBufferAllocator() {
    }

    public static ByteBuffer allocate(int size) {
        ByteBuffer directBuffer = ByteBuffer.allocateDirect(size);
        if (directBuffer == null) {
            if (size <= 0) {
                throw new IllegalArgumentException("Invalid allocation size: " + size);
            } else {
                throw new OutOfMemoryError("allocateDirectBuffer() returned null");
            }
        } else if (!directBuffer.isDirect()) {
            throw new AssertionError("allocateDirectBuffer() did not return a direct buffer");
        }

        return directBuffer;
    }

    public static ByteBuffer free(ByteBuffer buffer) {
        if (buffer == null) {
            return null;
        }

        if (!buffer.isDirect()) {
            throw new IllegalArgumentException("ByteBuffer must be direct");
        }

        return null;
    }

    public static ByteBuffer guardedAllocate(int size) {
        ByteBuffer buffer = null;
        try {
            buffer = allocate(size);
        } catch (OutOfMemoryError oomException) {
            return null;
        }
        return buffer;
    }
}

File: app/src/main/kotlin/org/libreoffice/kit/DirectBufferAllocator.kt
package org.libreoffice.kit

import java.nio.ByteBuffer

object DirectBufferAllocator {

    @Suppress("unused")
    private val LOGTAG: String = DirectBufferAllocator::class.java.simpleName

    fun allocate(size: Int): ByteBuffer {
        val directBuffer: ByteBuffer? = ByteBuffer.allocateDirect(size)
        if (directBuffer == null) {
            if (size <= 0) {
                throw IllegalArgumentException("Invalid allocation size: $size")
            } else {
                throw OutOfMemoryError("allocateDirectBuffer() returned null")
            }
        } else if (!directBuffer.isDirect) {
            throw AssertionError("allocateDirectBuffer() did not return a direct buffer")
        }
        return directBuffer
    }

    fun free(buffer: ByteBuffer?): ByteBuffer? {
        if (buffer == null) {
            return null
        }
        if (!buffer.isDirect) {
            throw IllegalArgumentException("ByteBuffer must be direct")
        }
        return null
    }

    fun guardedAllocate(size: Int): ByteBuffer? {
        return try {
            allocate(size)
        } catch (_: OutOfMemoryError) {
            null
        }
    }
}
